cityofaustin/atd-vz-data
# Replace Lambda function location_from_latlon with stored generated column

## Associated issues
Closes https://github.com/cityofaustin/atd-data-tech/issues/13055

~This PR recreates the `location_id` column but as a stored generated column for easier maintenance in keeping the column up to date whereas currently we rely on SQS and AWS lambda functions to do so. Related to Mikes work in https://github.com/cityofaustin/atd-vz-data/pull/1269 and I think we will have to coordinate deployment together to get this done most efficiently, bc like Mike talks about there are lots of views that depend on the `location_id` of cr3 crashes and noncr3 crashes.~

This PR now uses a trigger function to keep the location_id column up to date for the atd_txdot_crashes table.

## Testing
**URL to test:** 
Local db

**Steps to test:**

Using some of Mikes test/deployment steps here so shoutout Mike for coming up with great ways to test this!

1. Follow the deployment steps below in your local db instance
2. Test that the insert trigger works by running the following commands:

```plpgsql
-- insert a cr3 crash with coordinates that will associate to location_id 8337406736
INSERT INTO atd_txdot_crashes(latitude, longitude, crash_id)
VALUES (30.16841600, -97.78449000, 9999);

-- insert a cr3 crash with coordinates that will not associate to a location
INSERT INTO atd_txdot_crashes(latitude, longitude, crash_id)
VALUES (30.386136, -97.67309, 8888);

-- insert a cr3 crash without coordinates that will not associate to a location
INSERT INTO atd_txdot_crashes(latitude, longitude, crash_id)
VALUES (NULL, NULL, 7777);
``` 
3. Update an existing record with a location_id of NULL to have the `latitude_primary` and `longitude_primary` (these are the column values that are updated when a user uses the crash UI map to update coordinates) of a record with a location. After update, you should see the record associate with the same location as the record you took the lat/lon from.
4. Update an existing record with a location_id that is not null and update it with `latitude_primary` and `longitude_primary` coordinates from an existing record that has latitude and longitude and no location_id associated.


## Deployment steps

- [ ] Remove the Hasura trigger called location_from_latlon by going on to the Hasura console and go to the Events tab.
- [ ] Run SQL command to create the trigger function in [update_cr3_location.sql](https://github.com/cityofaustin/atd-vz-data/blob/dd674a85aca828307bf751784575822d06dbb263/atd-vzd/triggers/update_cr3_location.sql)
- [ ] Run SQL commands in the migrations file [migration_atd_txdot_crashes_2023_08_03--1127.sql](https://github.com/cityofaustin/atd-vz-data/blob/dd674a85aca828307bf751784575822d06dbb263/atd-vzd/migrations/migration_atd_txdot_crashes_2023_08_03--1127.sql)
- [ ] Switch off cr3 crash location housekeeping Airflow ETLs

---
#### Ship list
